### PR TITLE
Resource Isolation Feature: Extend CACHE SELECT to allow warming up backup replicas

### PR DESCRIPTION
**Pull request by anatoly2**
## Why I'm doing:

Currently a cache select statement meant to bring in 2 replicas of each tablet in some resource isolation groups might look like CACHE SELECT * FROM table PROPERTIES( "resource_isolation_groups"="writegroup, async", "num_replicas"="2");
This would intend to bring in each tablet into 2 CN in writegroup, and 2 CN in async. However, there may be some situation where we just want to bring in data for fallback purposes. For example,  say we have resource isolation "writegroup" which is responsible for data ingestion. As data is being ingested, it is automatically cached in the CN which is primarily responsible for each tablet. We don't need to do cache warmup for the primary replica because it's already cached, and it costs CPU to perform this noop. Instead, we'd like to issue a cache warmup such that we bring in backup replicas.

## What I'm doing:

I added the following syntax to support backup replicas: 
CACHE SELECT * FROM table PROPERTIES( "resource_isolation_groups"="writegroup", "num_backup_replicas"="1");
So this command will warmup the first backup replica. We could also specify num_backup_replicas=2 if we wanted 3 replicas total, etc. 

https://jira.pinadmin.com/browse/RTA-6821

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/system/TabletComputeNodeMapper.java [hunk 1]
```diff
@@ -203,21 +206,24 @@ public List<Long> computeNodesForTablet(Long tabletId) {
...
                 LOG.warn(String.format(
-                        "Requesting node for resource isolation group %s, to which" + " there is not a known CN assigned.",
+                        "Requesting node for resource isolation group %s, to which" +
+                                " there is not a known CN assigned.",
                         resourceIsolationGroup));
                 return Collections.emptyList();
             }
             TabletMap m = this.resourceIsolationGroupToTabletMapping.get(resourceIsolationGroup);
-            List<Long> computeNodes = m.tabletToComputeNodeId.get(tabletId, count);
+            List<Long> computeNodes = m.tabletToComputeNodeId.get(tabletId, count + skipCount).stream().skip(skipCount)
```
Comment (ctbrennan) [inline fe/fe-core/src/main/java/com/starrocks/system/TabletComputeNodeMapper.java:225]: Would you mind adding a unit test or two which tests `computeNodesForTablet` with a non-zero `skipCount`? I appreciate that you have a new test in CacheSelectBackendSelectorTest.java but I think it'd be ideal to have a smaller one at this layer as well.

File: fe/fe-core/src/main/java/com/starrocks/qe/CacheSelectComputeNodeSelectionProperties.java [hunk 2]
```diff
@@ -26,15 +25,19 @@
 public class CacheSelectComputeNodeSelectionProperties {
     public List<String> resourceIsolationGroups;
     public int numReplicasDesired;
+    public int numBackupReplicasDesired;
```
Comment (ctbrennan) [inline fe/fe-core/src/main/java/com/starrocks/qe/CacheSelectComputeNodeSelectionProperties.java:28]: This might be a good place to leave a comment describing the difference between the fields and noting that exactly one of `numReplicasDesired` and `numBackupReplicasDesired` should be nonzero.

File: fe/fe-core/src/main/java/com/starrocks/qe/FragmentScanRangeAssignment.java [hunk 3]
```diff
@@ -76,4 +76,4 @@ public String toDebugString() {
         }
         return sb.toString();
     }
-}
\ No newline at end of file
+}
```
Comment (ctbrennan) [inline fe/fe-core/src/main/java/com/starrocks/qe/FragmentScanRangeAssignment.java:79]: Can we remove this file from the PR?

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java [hunk 4]
```diff
@@ -154,28 +170,56 @@ public Void visitDataCacheSelectStatement(DataCacheSelectStatement statement, Co
...
-            statement.setNumReplicasDesired(Integer.parseInt(
-                    properties.getOrDefault("num_replicas", "1")));
-            if (statement.getNumReplicasDesired() < 1) {
-                throw new SemanticException("Num replicas must be positive");
+            Optional<Integer> maybeNumReplicas =
+                    Optional.ofNullable(properties.get(NUM_REPLICAS)).map(Integer::parseInt);
+            Optional<Integer> maybeNumBackupReplicas =
+                    Optional.ofNullable(properties.get(NUM_BACKUP_REPLICAS)).map(Integer::parseInt);
+
+            // accept only num_replicas or num_backup_replicas, not both
```
Comment (ctbrennan) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/DataCacheStmtAnalyzer.java:189]: All this logic looks good to me, but can we add a few test cases in `DataCacheStmtAnalyzerTest.testCacheSelectProperties` to ensure that we fail if NUM_REPLICAS and NUM_BACKUP_REPLICAS are both specified, etc. 

I'm actually noticing that a couple tests in that file are now failing.

**Review by ctbrennan — COMMENTED** [review]
Looks pretty good so far, left some comments.

**Review by saahilbarai — COMMENTED** [review]
Question for my understanding -

Suppose we have a write group of CN nodes. Like you mentioned in the PR description, that write group would have already had a primary replica cached upon ingestion. If we wanted to have two replicas without this PR we would have to perform a check against the hash ring to verify that for a given tablet id there was already one replica cached and that check is what we want to get rid of by exposing a secondary parameter "BackupReplicas". Essentially we now delegate more granular control of caching to the system admin to remove that check against the hash ring. 

If my understanding above is correct, how do I as a system admin know what combination of "DesiredReplicas" and "BackupReplicas" to issue when issuing a cache select statement. Is it mostly as simple as if the resource isolation group accepts writes then I exclusively modify "BackupReplicas"?
